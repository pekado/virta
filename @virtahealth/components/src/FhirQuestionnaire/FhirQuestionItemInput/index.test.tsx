import * as React from "react";
import { fireEvent } from "@testing-library/react-native";
import { createComponentWithVirtaContext } from "../../test";
import { FhirQuestionItemInput } from "./index";

describe("FHIR Question Item Input Field component", () => {
  const testSetQuestionnaireResponse = jest.fn((response) => response);

  afterEach(() => {
    testSetQuestionnaireResponse.mockClear();
  });

  it("renders input field for question item of type `question`", () => {
    const fhirQuestionInput = createComponentWithVirtaContext(
      <FhirQuestionItemInput
        itemData={{
          code: [{ code: "1234-5", display: "sample question" }],
          linkId: "test-linkId",
          type: "question",
        }}
        setQuestionnaireResponseItem={testSetQuestionnaireResponse}
      />
    );
    expect(fhirQuestionInput).toMatchSnapshot();
  });

  it("renders input field for question item that is not of type `question`", () => {
    const fhirQuestionInput = createComponentWithVirtaContext(
      <FhirQuestionItemInput
        itemData={{
          code: [{ code: "2345-6", display: "sample question in group" }],
          linkId: "test-decimal-linkId",
          extension: [{ valueCoding: { display: "mg/dL" }, url: "test/url" }],
          type: "decimal",
        }}
        setQuestionnaireResponseItem={testSetQuestionnaireResponse}
      />
    );
    expect(fhirQuestionInput).toMatchSnapshot();
  });

  it("will update questionnaire response with new answer item if one is not already present", () => {
    const expectedQuestionnaireResponseItem = {
      answer: [{ valueString: "this is some input" }],
      linkId: "test-random-linkId",
    };
    const questionItemInput = createComponentWithVirtaContext(
      <FhirQuestionItemInput
        itemData={{
          code: [{ code: "2345-6", display: "sample question in group" }],
          linkId: "test-random-linkId",
          type: "some-random-type",
        }}
        setQuestionnaireResponseItem={testSetQuestionnaireResponse}
      />
    );

    const textInput = questionItemInput.getByTestId("test-random-linkId");
    expect(textInput.props.value).toEqual("");

    fireEvent.changeText(textInput, "this is some input");

    expect(testSetQuestionnaireResponse).toHaveBeenCalledTimes(1);
    expect(testSetQuestionnaireResponse).toHaveBeenCalledWith(
      expectedQuestionnaireResponseItem
    );
  });

  it("will update existing answer in questionnaire response if it is already present", () => {
    const testQuestionnaireResponseItem = {
      answer: [{ valueDecimal: 0.2 }],
      linkId: "test-decimal-linkId",
    };

    const expectedQuestionnaireResponse = {
      answer: [{ valueDecimal: "100.1" }],
      linkId: "test-decimal-linkId",
    };
    const fhirQuestionnaireItem = createComponentWithVirtaContext(
      <FhirQuestionItemInput
        itemData={{
          code: [
            { code: "6543-2", display: "decimal sample question in group" },
          ],
          linkId: "test-decimal-linkId",
          extension: [{ valueCoding: { display: "mg/dL" }, url: "test/url" }],
          type: "decimal",
        }}
        setQuestionnaireResponseItem={testSetQuestionnaireResponse}
        questionnaireResponseItem={testQuestionnaireResponseItem}
      />
    );

    const textInput = fhirQuestionnaireItem.getByTestId("test-decimal-linkId");
    expect(textInput.props.value).toEqual("0.2");

    fireEvent.changeText(textInput, "100.1");

    expect(testSetQuestionnaireResponse).toHaveBeenCalledTimes(1);
    expect(testSetQuestionnaireResponse).toHaveBeenCalledWith(
      expectedQuestionnaireResponse
    );
  });

  // This scenario happens for certain lab tests that need to accept `>` and `<` inputs,
  // such as '< 2.4' or '> 4.3'
  it("will allow string answers for numeric questions that have a regex", () => {
    const testQuestionnaireResponseItem = {
      answer: [{ valueDecimal: 0.2 }],
      linkId: "test-decimal-linkId",
    };
    const expectedQuestionnaireResponse = {
      answer: [{ valueDecimal: "> 0.2" } as any],
      linkId: "test-decimal-linkId",
    };
    const fhirQuestionnaireItem = createComponentWithVirtaContext(
      <FhirQuestionItemInput
        itemData={{
          code: [
            { code: "6543-2", display: "decimal sample question in group" },
          ],
          linkId: "test-decimal-linkId",
          extension: [
            { valueCoding: { display: "mg/dL" }, url: "test/url" },
            {
              url: "http://hl7.org/fhir/StructureDefinition/regex",
              valueString: "^[><]\\s?\\d+(\\.\\d+)?$",
            },
          ],
          type: "decimal",
        }}
        setQuestionnaireResponseItem={testSetQuestionnaireResponse}
        questionnaireResponseItem={testQuestionnaireResponseItem}
      />
    );

    const textInput = fhirQuestionnaireItem.getByTestId("test-decimal-linkId");
    expect(textInput.props.value).toEqual("0.2");

    fireEvent.changeText(textInput, "> 0.2");
    expect(testSetQuestionnaireResponse).toHaveBeenCalledTimes(1);
    expect(testSetQuestionnaireResponse).toHaveBeenCalledWith(
      expectedQuestionnaireResponse
    );
  });

  it("will map `quantity` type answers to `valueQuantity`", () => {
    const testQuestionnaireResponseItem = {
      answer: [{ valueQuantity: { value: 0.2 } }],
      linkId: "test-quantity-linkId",
    };
    const expectedQuestionnaireResponse = {
      answer: [{ valueQuantity: { value: "20" } } as any],
      linkId: "test-quantity-linkId",
    };
    const fhirQuestionnaireItem = createComponentWithVirtaContext(
      <FhirQuestionItemInput
        itemData={{
          code: [
            { code: "6543-2", display: "decimal sample question in group" },
          ],
          linkId: "test-quantity-linkId",
          extension: [{ valueCoding: { display: "mg/dL" }, url: "test/url" }],
          type: "quantity",
        }}
        setQuestionnaireResponseItem={testSetQuestionnaireResponse}
        questionnaireResponseItem={testQuestionnaireResponseItem}
      />
    );

    const textInput = fhirQuestionnaireItem.getByTestId("test-quantity-linkId");
    expect(textInput.props.value).toEqual("0.2");

    fireEvent.changeText(textInput, "20");
    expect(testSetQuestionnaireResponse).toHaveBeenCalledTimes(1);
    expect(testSetQuestionnaireResponse).toHaveBeenCalledWith(
      expectedQuestionnaireResponse
    );
  });
});
