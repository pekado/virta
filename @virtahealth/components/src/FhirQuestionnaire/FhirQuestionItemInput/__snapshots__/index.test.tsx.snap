// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FHIR Question Item Input Field component renders input field for question item of type \`question\` 1`] = `
<View
  style={
    Array [
      Object {
        "marginRight": 32,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexBasis": "auto",
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 0,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "marginRight": 0,
            "marginTop": 0,
            "minWidth": 205,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "rgb(36, 38, 43)",
              "fontFamily": "Whitney-Book",
              "fontSize": 12,
              "fontWeight": "500",
              "letterSpacing": 0,
              "lineHeight": 12,
            },
          ]
        }
      >
        sample question
      </Text>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            },
            Object {
              "display": "flex",
              "marginTop": 10,
            },
          ]
        }
      >
        <View
          editable={true}
          isFocused={false}
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "flex-start",
              },
              Object {
                "alignItems": "center",
                "backgroundColor": "rgb(255,255,255)",
                "borderColor": "rgb(214, 217, 221)",
                "borderRadius": 6,
                "borderStyle": "solid",
                "borderWidth": 2,
                "display": "flex",
                "flexBasis": "auto",
                "flexGrow": 1,
                "flexShrink": 0,
              },
            ]
          }
        >
          <TextInput
            editable={true}
            isFocused={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "rgb(0,0,0)",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "fontFamily": "Whitney-Book",
                  "fontSize": 20,
                  "minHeight": 40,
                  "paddingBottom": 8,
                  "paddingLeft": 12,
                  "paddingRight": 12,
                  "paddingTop": 8,
                },
              ]
            }
            testID="test-linkId"
            value=""
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "height": 36,
        },
      ]
    }
  />
</View>
`;

exports[`FHIR Question Item Input Field component renders input field for question item that is not of type \`question\` 1`] = `
<View
  style={
    Array [
      Object {
        "marginRight": 32,
      },
    ]
  }
>
  <View
    style={
      Array [
        Object {
          "display": "flex",
          "flexBasis": "auto",
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 0,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "marginRight": 0,
            "marginTop": 0,
            "minWidth": 205,
          },
        ]
      }
    >
      <Text
        style={
          Array [
            Object {
              "color": "rgb(36, 38, 43)",
              "fontFamily": "Whitney-Book",
              "fontSize": 12,
              "fontWeight": "500",
              "letterSpacing": 0,
              "lineHeight": 12,
            },
          ]
        }
      >
        sample question in group
      </Text>
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "flexDirection": "row",
              "justifyContent": "flex-start",
            },
            Object {
              "display": "flex",
              "marginTop": 10,
            },
          ]
        }
      >
        <View
          editable={true}
          isFocused={false}
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "flex-start",
              },
              Object {
                "alignItems": "center",
                "backgroundColor": "rgb(255,255,255)",
                "borderColor": "rgb(214, 217, 221)",
                "borderRadius": 6,
                "borderStyle": "solid",
                "borderWidth": 2,
                "display": "flex",
                "flexBasis": "auto",
                "flexGrow": 1,
                "flexShrink": 0,
              },
            ]
          }
        >
          <TextInput
            editable={true}
            isFocused={false}
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            style={
              Array [
                Object {
                  "backgroundColor": "transparent",
                  "color": "rgb(0,0,0)",
                  "flexBasis": 0,
                  "flexGrow": 1,
                  "flexShrink": 1,
                  "fontFamily": "Whitney-Book",
                  "fontSize": 20,
                  "minHeight": 40,
                  "paddingBottom": 8,
                  "paddingLeft": 12,
                  "paddingRight": 12,
                  "paddingTop": 8,
                },
              ]
            }
            testID="test-decimal-linkId"
            value=""
          />
          <Text
            numberOfLines={1}
            style={
              Array [
                Object {
                  "color": "rgb(36, 38, 43)",
                  "fontFamily": "Whitney-Book",
                  "fontSize": 20,
                  "fontWeight": "500",
                  "letterSpacing": 0,
                  "lineHeight": 20,
                  "textDecorationColor": "black",
                  "textDecorationLine": "none",
                  "textDecorationStyle": "solid",
                },
                Object {
                  "alignItems": "center",
                  "color": "rgb(89, 94, 101)",
                  "display": "flex",
                  "flexBasis": "auto",
                  "fontFamily": "Whitney-Book",
                  "fontSize": 14,
                  "fontWeight": "300",
                  "letterSpacing": 0,
                  "lineHeight": 18,
                  "minHeight": 40,
                  "paddingBottom": 8,
                  "paddingLeft": 12,
                  "paddingRight": 12,
                  "paddingTop": 8,
                },
              ]
            }
          >
            mg/dL
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "height": 36,
        },
      ]
    }
  />
</View>
`;
